scio-core/src/main/scala/com/spotify/scio/nio/ScioIO.scala

```diff
@@ -42,5 +42,36 @@ trait ScioIO[T] {
   def tap(read: ReadP): Tap[T]
```

Comment:
@nevillelyh Should this method return a `Future[Tap[T]]` instead of simply `Tap[T]` ?
The existing methods (`textFile`, `protobufFile`, `objectFile`, `avroFile` and `typedAvroFile`) from `Taps` are all based on `Taps.mkTap` which returns a `Future`. I'm assuming they return a Future to support polling.

I also noticed that the implementation of `TextIO` is copied from `FileStorage`, `ScioContext` and `SCollection`.  It's probably not great to duplicate that much code. Especially code from `FileStorage` since this class is private and there's no reason not to just delegate to it.

Reply:
@jto
regarding `Tap[T]`, only write method should return the `Future[Tap[T]]` not this method hence why we have `write(data: SCollection[T], params: WriteP): Future[Tap[T]]`, this method is to create a tap which we can use inside write method. write correct me if I am wrong. 

 The reason behind duplicating code from `FileStorage` is to remove that dependency from new `TextIO` and later delete(or clean all un-necessary util classes) that class.  We can see how much code we need to share with other new `IO` classes and restructure the code accordingly. Following the rule, premature optimization is bad practice :).

Reply:
fair enough

Reply:
Hum actually @syodage , this can't be correct.
`ScioIO.tap` needs to be passed a ReadP which `ScioIO.write` does not have.